src/utils/hooks/useGetResponsesAndElementState.tsx

```diff
@@ -0,0 +1,127 @@
+
+    const templateSections = data?.applicationBySerial?.template?.templateSections?.nodes
+
+    const templateElements: TemplateElementState[] = []
+
+    templateSections?.forEach((sectionNode) => {
+      sectionNode?.templateElementsBySectionId?.nodes.forEach((element) => {
+        if (element) templateElements.push(element as TemplateElementState) // No idea why ...[spread] doesn't work here.
+      })
+    })
+
+    const currentResponses = {} as ResponsesByCode
+    const currentFullResponses = {} as ResponsesFullByCode
+
+    if (applicationResponses) {
+      applicationResponses.forEach((response) => {
+        const code = response?.templateElement?.code
+        if (code) {
+          currentResponses[code] = response?.value?.text
+          currentFullResponses[code] = response?.value
+        }
+      })
+    }
+
+    setResponsesByCode(currentResponses)
+    setResponsesFullByCode(currentFullResponses)
+    setElementsExpressions(templateElements)
+    setLoading(false)
+  }, [data, apolloError])
+
+  useEffect(() => {
+    evaluateElementExpressions().then((result) => setElementsState(result))
+  }, [responsesByCode])
+
+  async function evaluateElementExpressions() {
+    const promiseArray: Promise<EvaluatedElement>[] = []
+    elementsExpressions.forEach((element) => {
+      promiseArray.push(evaluateSingleElement(element))
+    })
+    const evaluatedElements = await Promise.all(promiseArray)
+    const elementsState: any = {}
+    evaluatedElements.forEach((element) => {
+      elementsState[element.code] = {
+        id: element.id,
+        category: element.category,
+        isEditable: element.isEditable,
+        isRequired: element.isRequired,
+        isVisible: element.isVisible,
+        // isValid: element.isValid,
+      }
+    })
+    return elementsState
+  }
+
+  async function evaluateSingleElement(element: any): Promise<EvaluatedElement> {
+    const evaluationParameters = {
+      objects: [responsesByCode, responsesFullByCode], // TO-DO: Also send user/org objects etc.
+      // graphQLConnection: TO-DO
+    }
+    const isEditable = evaluateExpression(element.isEditable)
+    const isRequired = evaluateExpression(element.isRequired)
+    const isVisible = evaluateExpression(element.visibilityCondition)
+    // const isValid = evaluateExpression(element.validation)
+    const results = await Promise.all([isEditable, isRequired, isVisible])
+    const evaluatedElement = {
+      code: element.code,
+      id: element.id,
+      category: element.category,
+      isEditable: results[0] as boolean,
+      isRequired: results[1] as boolean,
+      isVisible: results[2] as boolean,
+      // isValid: results[3],
+    }
+    return evaluatedElement
+  }
+
+  return {
+    error,
+    loading,
+    responsesByCode,
+    responsesFullByCode,
+    elementsState,
+  }
+}
+
+function checkForApplicationErrors(data: GetElementsAndResponsesQuery | undefined) {
+  if (data?.applicationBySerial) {
```

Comment:
Check if the result `data` isn't undefined and if `data.applicationBySerial` isn't undefined it's better, so you don't need to be checking inside the scope of the useEffect - since that's called after doing this checks

Reply:
Oh and `data.applicationBySerial.applicationResponses` just in case

Reply:
Oops needs to also check for correct responses for: `template` and `template?.templateSections`

Reply:
Okay, hope what I've done is okay now:
```
function checkForApplicationErrors(data: GetElementsAndResponsesQuery | undefined) {
  if (
    !data?.applicationBySerial?.applicationResponses?.nodes ||
    !data?.applicationBySerial?.template?.templateSections
  )
    return 'Data undefined'
  if (data?.applicationBySerial?.applicationResponses?.nodes.length === 0)
    return 'No responses found'
  return null
}
```